### PR TITLE
Vmap ensemble training and jaxmd inference

### PR DESCRIPTION
**Pull request by M-R-Schaefer**
This PR implements the vmapped training of multiple models at once.
These models can be loaded into the ASE calc for ensembling similarly to training multiple models separately. 
Furthermore, ensembles can now be used in jaxmd.
However, uncertainties are not logged yet.

### CODE REVIEW

File: apax/bal/api.py [hunk 1]
```diff
@@ -82,7 +82,9 @@ def kernel_selection(
     config, params = restore_parameters(model_dir)
 
     n_train = len(train_atoms)
-    dataset = initialize_dataset(config, RawDataset(atoms_list=train_atoms + pool_atoms))
+    dataset = initialize_dataset(
+        config, RawDataset(atoms_list=train_atoms + pool_atoms), False
+    )
```
Comment (PythonFZ) [inline apax/bal/api.py:87]: The `config` and `RawDataset` are clear to me, but passing just `False` seems a bit odd - I'd suggest using a keyword argument instead.

File: apax/md/nvt.py [hunk 2]
```diff
@@ -19,12 +19,36 @@
...
+    if n_models > 1:
+        energy_fn = ensemble
+    else:
+        energy_fn = model
+
+    energy_fn = partial(
+        energy_fn,
+        params,
+        Z=numbers,
+        box=box,  # TODO IS THIS CORRECT FOR NPT???
```
Comment (PythonFZ) [inline apax/md/nvt.py:45]: > # TODO IS THIS CORRECT FOR NPT???

I don't know, but if you want to merge this, one should certainly raise an issue that this must be checked!

File: apax/train/checkpoints.py [hunk 3]
```diff
@@ -14,13 +14,56 @@
 log = logging.getLogger(__name__)
 
 
-def load_state(model, params, tx, ckpt_dir):
+def check_for_ensemble(params):
+    flat_params = flatten_dict(params)
+    shapes = [v.shape[0] for v in flat_params.values()]
+    is_ensemble = shapes == shapes[::-1]
```
Comment (PythonFZ) [inline apax/train/checkpoints.py:20]: Why is it an ensemble if the shape is equal to its reversed?

File: apax/train/checkpoints.py [hunk 4]
```diff
@@ -14,13 +14,56 @@
...
+def create_train_state(model, params, tx):
+    n_models = check_for_ensemble(params)
+
+    def inner(params):
+        state = train_state.TrainState.create(
+            apply_fn=model,
+            params=params,
+            tx=tx,
+        )
+        return state
```
Comment (PythonFZ) [inline apax/train/checkpoints.py:37]: I don't like the function name, could it be `compute_inner` ? `inner` sounds more like a variable.

File: apax/train/checkpoints.py [hunk 5]
```diff
@@ -14,13 +14,56 @@
...
+    keys = jax.random.split(rng_key, num=n_models + 1)
+    rng_key, model_rng = keys[0], keys[1:]
+
+    log.info(f"initializing {n_models} models")
+
+    if n_models == 1:
+        params = model.init(model_rng[0], *sample_input)
+    elif n_models > 1:
+        num_args = len(sample_input)
+        in_axes = (0, *[None] * num_args)
```
Comment (PythonFZ) [inline apax/train/checkpoints.py:55]: :sweat_smile:

File: apax/train/trainer.py [hunk 6]
```diff
@@ -156,38 +158,76 @@ def calc_loss(params, inputs, labels, loss_fn, model):
...
+def make_ensemble_update(update_fn: Callable) -> Callable:
+    v_update_fn = jax.vmap(update_fn, (0, None, None), (0, 0, 0))
+
+    def ensemble_update_fn(state, inputs, labels):
+        loss, predictions, state = v_update_fn(state, inputs, labels)
+
+        mean_predictions = jax.tree_map(lambda x: jnp.mean(x, axis=0), predictions)
+        mean_loss = jnp.mean(loss)
+        # TODO Add std to predictions
+        return mean_loss, mean_predictions, state
```
Comment (PythonFZ) [inline apax/train/trainer.py:170]: could this function be moved outside and returned with `partial(ensemble_update_fn, v_update_fn=v_update_fn)` to reduce code cuplication (see `make_ensemble_eval`) ?

File: apax/train/trainer.py [hunk 7]
```diff
@@ -156,38 +158,76 @@ def calc_loss(params, inputs, labels, loss_fn, model):
...
+    def update_step(state, inputs, labels):
         (loss, predictions), grads = grad_fn(state.params, inputs, labels)
 
         if rho > 1e-6:
             # SAM step
             grad_norm = global_norm(grads)
             eps = jax.tree_map(lambda g, n: g * rho / n, grads, grad_norm)
             params_eps = jax.tree_map(lambda p, e: p + e, state.params, eps)
-            (loss, _), grads = grad_fn(params_eps, inputs, labels)
+            (loss, _), grads = grad_fn(params_eps, inputs, labels)  # maybe get rid of SAM
```
Comment (PythonFZ) [inline apax/train/trainer.py:201]: If it is important, please raise an issue before merging the PR

**Review by PythonFZ — COMMENTED** [review]
Tried my best. Is there a good way to document / keep track of all the dimensions you are passing around everywhere? Maybe some docstrings that help with the inputs and outputs of the model for better maintainability?

File: apax/train/checkpoints.py (hunk 3 above)
Comment (M-R-Schaefer) [reply apax/train/checkpoints.py:20]: I suppose a clearer way of checking would be to see if all the first dimensions are the same

File: apax/train/trainer.py (hunk 6 above)
Comment (M-R-Schaefer) [reply apax/train/trainer.py:170]: The only part that is really duplicated is
```
mean_predictions = jax.tree_map(lambda x: jnp.mean(x, axis=0), predictions)
mean_loss = jnp.mean(loss)
```

I think 2 lines are acceptable

Comment (M-R-Schaefer) [reply apax/train/trainer.py:170]: I guess I could unify them by using tuple (un)packing, but I'm not sure if this helps readability.
We should however adopt proper type hints. I'll make an issue.

**Pull request merged**